src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\RedisExt;
+
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+use Symfony\Component\Messenger\Exception\LogicException;
+
+/**
+ * A Redis connection.
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ * @author Alexander Schranz <alexander@sulu.io>
+ *
+ * @final
+ *
+ * @experimental in 4.3
+ */
+class Connection
+{
+    private $connection;
+    private $stream;
+    private $group;
+    private $consumer;
+    private $blockingTimeout;
+
+    public function __construct(array $configuration, array $connectionCredentials = [], array $redisOptions = [])
+    {
+        $this->connection = new \Redis();
+        $this->connection->connect($connectionCredentials['host'] ?? '127.0.0.1', $connectionCredentials['port'] ?? 6379);
+        $this->connection->setOption(\Redis::OPT_SERIALIZER, $redisOptions['serializer'] ?? \Redis::SERIALIZER_PHP);
+        $this->stream = $configuration['stream'] ?? 'messages';
+        $this->group = $configuration['group'] ?? 'symfony';
+        $this->consumer = $configuration['consumer'] ?? 'consumer';
+        $this->blockingTimeout = $redisOptions['blocking_timeout'] ?? null;
+    }
+
+    public static function fromDsn(string $dsn, array $redisOptions = []): self
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The given Redis DSN "%s" is invalid.', $dsn));
+        }
+
+        $pathParts = explode('/', $parsedUrl['path']);
+
+        $stream = $pathParts[1] ?? '';
+        $group = $pathParts[2] ?? '';
+        $consumer = $pathParts[3] ?? '';
+
+        $connectionCredentials = [
+            'host' => $parsedUrl['host'] ?? '127.0.0.1',
+            'port' => $parsedUrl['port'] ?? 6379,
+        ];
+
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $parsedQuery);
+            $redisOptions = array_replace_recursive($redisOptions, $parsedQuery);
+        }
+
+        return new self(['stream' => $stream, 'group' => $group, 'consumer' => $consumer], $connectionCredentials, $redisOptions);
+    }
+
+    public function get(): iterable
+    {
+        $messages = $this->connection->xreadgroup(
+            $this->group,
+            $this->consumer,
+            [$this->stream => '>'],
+            1,
+            $this->blockingTimeout
+        );
+
+        if (false === $messages) {
+            throw new LogicException(
+                $this->connection->getLastError() ?: 'Unexpected redis stream error happened.'
+            );
+        }
+
+        foreach ($messages[$this->stream] as $key => $message) {
```

Comment:
What if you crash (segafault) before ack/nack the message? what happens to the message flagged as "pending" by redis,

Reply:
Thx for the review added to the checklist. I think we need first receive all pending messages first when a consumer crash or is restarted.

Reply:
As I could not found that Redis is adding pending messages later back to the stream after a timeout. I would think we need read first the pending messages again which where not finished and then later the new messages?

Reply:
From the blog post: https://redis.io/topics/streams-intro

> As you can see the idea here is to start consuming the history, that is, our list of pending messages. This is useful because the consumer may have crashed before, so in the event of a restart we want to read again messages that were delivered to us without getting acknowledged. This way we can process a message multiple times or one time (at least in the case of consumers failures, but there are also the limits of Redis persistence and replication involved, see the specific section about this topic).


You also have XPENDING command for that.

Reply:
@jderusso thank you. xpending give just a list of pending messages with ids but you can give `0` instead of `>` to xreadgroup get only the current pending messages which I now use.

> Any other ID, that is, 0 or any other valid ID or incomplete ID (just the millisecond time part), will have the effect of returning entries that are pending for the consumer sending the command. So basically if the ID is not >, then the command will just let the client access its pending entries: delivered to it, but not yet acknowledged.

https://redis.io/commands/xreadgroup